src/lib/QueryParser.js

```diff
@@ -0,0 +1,459 @@
+/**
+ * @license
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+foam.CLASS({
+  package: 'foam.parse',
+  name: 'QueryParser',
+  axioms: [foam.pattern.Singleton.create()],
```

Comment:
Alternative design idea.

What if we make this a multiton based upon the id of the class you pass in?

Then we can just do foam.parse.QueryParser.create({ of: 'some.class' }), in every spot we need the query parser, no need to pass it around or make sure we only call parserForCls() once per class.

Then QueryParser could extend Grammar and could be used like this.

this.QueryParser.create({ of: 'some.Class' }).parseString(query)